xep-0384.xml

```diff
@@ -514,11 +530,12 @@
     </section3>
     <section3 topic='Message structure description' anchor='message-structure-description'>
       <p>
-        An OMEMO encrypted message is specified to include an &lt;encrypted&gt; element in the <tt>&ns;</tt> namespace. It always contains two child nodes, the &lt;header&gt; and the &payload; element.
-        The &lt;header&gt; element has an attribute named 'sid' referencing the device id of the sending device and contains one or multiple &lt;keys&gt; elements, each with an attribute 'jid' of one of the recipients bare JIDs as well as one or multiple &lt;key&gt; elements.
+        An OMEMO encrypted message is specified to include an &lt;encrypted&gt; element in the <tt>&ns;</tt> namespace. It contains up to two child nodes, the &lt;header&gt; and the &payload; element. The &lt;header&gt; element must always be present, the &payload; element must be present unless an empty message is sent, as described below.
+        The &lt;header&gt; element has an attribute named 'sid' referencing the device id of the sending device and contains one or multiple &lt;keys&gt; elements, each with an attribute 'jid' of one of the recipients bare JIDs, as well as one or multiple &lt;key&gt; elements.
         A &lt;key&gt; element has an attribute named 'rid' referencing the device id of the recipient device, and an attribute named 'kex' which defaults to 'false' and indicates if the enclosed encrypted message includes a key exchange. The key and HMAC encrypted using the long-standing OMEMO session for that recipient device are encoded using base64 and placed as text content into the &lt;key&gt; element.
         The encrypted &content; element is encoded using base64 and placed as text content into the &payload; element.
       </p>
+      <p>A special case are <em>empty</em> messages, which are used in various places throughout the protocol purely to manage sessions and not to transfer content. With empty messages, the step of creating and encrypting the &payload; element is skipped. Instead of encrypting the key and authentication tag of the &payload; ciphertext with the Double Ratchet session, 32 zero-bytes are encrypted with the Double Ratchet session directly. The resulting OMEMOKeyExchange or OMEMOAuthenticatedMessage are put into &lt;key&gt; elements as usual, but the &payload; element is omitted altogether, so that the &lt;encrypted&gt; element only contains a &lt;header&gt;.</p>
```

Comment:
Why do we use zeros here? Wouldn't that allow for potential known plaintext attacks?

Reply:
First of all, if you're not working on something super niche, you'll probably never get in touch with a cipher that is (known to be) vulnerable to known plaintext attacks. And even if AES-CBC was vulnerable, all you could find is the key for a single message, which, due to OMEMO's ratcheting, is useless except for that single message, which is known anyway :)

In summary it doesn't really matter what we encrypt here, except that _nothing_ is not an option. And because we need _something_, I felt like fixed zeros for an "empty" message are somewhat clearer/more intuitive than e.g. 32 random bytes. But that's purely subjective I guess.

Reply:
> And even if AES-CBC was vulnerable, all you could find is the key for a single message, which, due to OMEMO's ratcheting, is useless except for that single message, which is known anyway :)

Not sure about that. Since the key for the next message going the same direction is directly derived from the key an attacker would have discovered, they'd potentially gain access to more than just one message.
I'm not an expert in crypto, so I can't tell how likely it is that a known-plaintext attack will affect AES, but it also wouldn't hurt to simply encrypt 32 bytes of randomness, would it?

Reply:
No, the next key is not directly derived from the previous one. One step of the ratchet derives 32 bytes, 16 of which are used for the next derivation, the other 16 are used as the key. So having the key doesn't tell anything about the other 16 that are kept in secret for the next step.

Reply:
(sorry, it's 2x32 not 16)

Reply:
Ah, great! :)